test/unit.coffee

```diff
@@ -0,0 +1,382 @@
+    connection_close_error = new Error "Connection closed"
+    initialized_thrift_service =
+      fn: (arg, cb) =>
+        setImmediate => @mock_connection.emit "close"
+        wait = -> cb null, arg
+        setTimeout wait, 1000
+    thrift =
+      createConnection: =>
+        setImmediate => @mock_connection.emit "connect"
+        @mock_connection
+      createClient: sinon.stub().returns initialized_thrift_service
+    wrappedPool = thriftPool thrift, @thriftService, {"host", "port"}, {timeout: 100}
+    wrappedPool.fn "abc", (err, data) ->
+      assert.deepEqual err, connection_close_error
+      assert.equal data, null
+      done()
+
+
+# Create_pool unit makes sure create_pool properly initializes a generic-pool
+# for thrift. Each of these tests creates a new pool which uses a single mock
+# connection (generic_pool handles and tests all other pooling functions).
+#
+# The following is tested:
+#   - "create", "destroy", and "validate" functions are initialized properly
+#     and behave as expected.
+#   - Emitting thrift events - "connect", "error", and "close", yield proper
+#     behavior for a connection when it is in a pool, being created, or acquired.
+#   - Passing "timeout" in thrift_options adds "timeout" listener and behaves
+#     as expected.
+describe 'create_pool unit', ->
+  # Initialize mocks to use in pool creation
+  before ->
+    @mock_connection = connection_mock()
+    @thrift =
+      createConnection: sinon.stub().returns @mock_connection
+    @options =
+      host: "host"
+      port: "port"
+      max_connections: 20
+      min_connections: 0
+      idle_timeout: 30000
+
+  # Function helpers
+  before ->
+    # Makes sure err and connection match that of a successful acquire.
+    @assert_valid = (err, connection) ->
+      assert.ifError err
+      assert.notEqual connection, null
+      assert.equal connection.__ended, false
+
+    # Makes sure that calling acquire on connection destroys
+    # that connection and creates and returns another one.
+    @acquire_destroys = (pool, cb) ->
+      setImmediate => @mock_connection.emit "connect"
+      pool.acquire (err, connection) =>
+        @assert_valid err, connection
+        assert @mock_connection.end.called
+        assert @thrift.createConnection.called
+        pool.release connection
+        cb()
+
+    # Emits a connect event, acquires, and releases the connection.
+    @connect_release = (pool, cb) =>
+      # Connection succeeds in creation and is released
+      setImmediate => @mock_connection.emit "connect"
+      pool.acquire (err, connection) =>
+        @assert_valid err, connection
+        pool.release connection
+        cb()
+
+    # Used to test the event emitters that invalidate a connection.
+    # Flow: creates/releases connection -> emits emit_fn -> acquires connection.
+    @emit_invalidates_connection = (emit_fn, done, thrift_options = {}) =>
```

Comment:
also, does this need to be defined in the before? or can we just define it anywhere?